landa/utils.py

```diff
@@ -78,3 +81,160 @@ def autocommit():
 	frappe.db.auto_commit_on_many_writes = True
 	yield
 	frappe.db.auto_commit_on_many_writes = flag_value
+
+
+def remove_from_table(table_doctype: str, link_field: str, value: str):
+	table = frappe.qb.DocType(table_doctype)
+	query = (
+		frappe.qb.from_(table)
+		.select(table.parenttype, table.parent, table.parentfield, table.idx)
+		.where(table[link_field] == value)
+		.distinct()
+	)
+
+	for parent_type, parent_name, parent_field, idx in query.run():
+		pop_from_table(parent_type, parent_name, parent_field, idx)
+
+
+def pop_from_table(parent_type, parent_name, parent_field, idx):
+	doc = frappe.get_doc(parent_type, parent_name)
+	doc.get(parent_field).pop(idx - 1)
+	doc.save(ignore_permissions=True)
+
+
+def delete_dynamically_linked(
+	doctype: str, linked_doctype: str, linked_name: str
+) -> None:
+	"""Delete all records of `doctype` that are linked to `linked_name` of `linked_doctype` via a Dynamic Link table."""
+	dl = frappe.qb.DocType("Dynamic Link")
+	for name, parent_field, idx in (
+		frappe.qb.from_(dl)
+		.select(dl.parent, dl.parentfield, dl.idx)
+		.where(
+			(dl.parenttype == doctype)
+			& (dl.link_doctype == linked_doctype)
+			& (dl.link_name == linked_name)
+		)
+		.run()
+	):
+		doc: Document = frappe.get_doc(doctype, name)
+		if any(
+			row
+			for row in doc.links
+			if row.link_doctype == linked_doctype and row.link_name != linked_name
+		):
+			# if there are links to records of the same doctype, for example, an
+			# address belonging to multiple members
+			pop_from_table(doctype, name, parent_field, idx)
+			continue
+
+		frappe.delete_doc(
+			doctype, doc.name, delete_permanently=True, ignore_permissions=True
+		)
+
+
+def get_link_fields(doctype: str, as_dict: int = 1):
+	# TODO: handle fields changed by Property Setter(?)
+	return frappe.db.sql(
+		"""
+		SELECT
+			parent,
+			fieldname,
+			reqd,
+			(SELECT issingle FROM tabDocType dt WHERE dt.name = df.parent) AS is_in_single,
+			(SELECT istable FROM tabDocType dt WHERE dt.name = df.parent) AS is_in_table
+		FROM tabDocField df
+		WHERE
+			df.options=%(doctype)s AND df.fieldtype='Link'
+
+		UNION DISTINCT
+
+		SELECT
+			dt AS parent,
+			fieldname,
+			reqd,
+			(SELECT issingle FROM tabDocType dt WHERE dt.name = df.dt) AS is_in_single,
+			(SELECT istable FROM tabDocType dt WHERE dt.name = df.dt) AS is_in_table
+		FROM `tabCustom Field` df
+		WHERE
+			df.options=%(doctype)s AND df.fieldtype='Link'
+		""",
+		{"doctype": doctype},
+		as_dict=as_dict,
+	)
+
+
+def purge_all(doctype: str, name: str) -> None:
```

Comment:
```suggestion
def delete_all_links(doctype: str, name: str) -> None:
```
More explicit and has similar context to the **Links** button in the document menu that shows linked documents